### DISCUSSION TITLE
Port is already in use after closing and opening ntrip caster

### DISCUSSION THREAD

**Discussion by karlranseierausrom** [semuconsulting/PyGPSClient#106] [General]
When I connect the NTRIP Caster, disconnect and reconnect it, it says "Error starting socket server [ERRNO 98] address already in use". Yes, it says socket server when trying to start NTRIP caster.

When changing the config from Socket server to NTRIP Caster it is loading the default (?) port 2101 or at least the one saved in the config. When changing to socket server it stays at 2101. 

edit: 
'Pi@RTKbase:~ $ fuser 2101/tcp
2101/tcp:            17098
Pi@RTKbase:~ $ ls -l /proc/17098/exe
lrwxrwxrwx 1 Pi Pi 0 Jan 14 20:42 /proc/17098/exe -> /usr/bin/python3.9'

So it used by python and this should be PyGPSClient. I have no idea how to get more information.

**Comment by semuadmin** [comment]
Hi @karlranseierausrom

This behaviour is operating system dependent to a large extent. Depending on how the socket connection is terminated, the operating system 'keeps hold' of the socket port for a short period before releasing it for reuse - typically a few minutes.

I recommend that, as far as is practicable, you shut down any client(s) _first_, _then_ close the PyGPSClient socker server/caster by unchecking the 'Socket Server/NTRIP Caster' checkbox _before_ closing PyGPSClient. If this sequence is followed, the port should be immediately reusable under normal circumstances.

There are ways to address the `address already in use` in Python code (e.g. use of SO_REUSEADDR, SO_REUSEPORT and/or SO_LNGER flags) but none are completely foolproof. When I get the time I may look into improving socket server handling in pygnssutils (which is used by PyGPSClient), but for the time being I recommend you follow the procedure above or simply wait a few minutes for the socket port to be released by the operating system.
